File: Java/src/org/hpms/mw/distcrud/Dispatcher.java
package org.hpms.mw.distcrud;

import java.io.IOException;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

import org.hpms.mw.distcrud.IRequired.CallMode;

class Dispatcher implements IDispatcher, IProtocol {

   private static class Operation {

      final IOperation _operation;
      final Arguments  _arguments;
      final String     _intrfcName;
      final String     _opName;
      final int        _callId;

      public Operation(
         IOperation operation,
         Arguments  arguments,
         String     intrfcName,
         String     opName,
         int        callId )
      {
         _operation  = operation;
         _arguments  = arguments;
         _intrfcName = intrfcName;
         _opName     = opName;
         _callId     = callId;
      }
   }

   final static class ProvidedImpl implements IProvided {

      private final Map<String, IOperation> _opsInOut = new HashMap<>();

      @Override
      public boolean addOperation( String opName, IOperation executor ) {
         return _opsInOut.put( opName, executor ) == null;
      }
   }

   private final ParticipantImpl           _participant;
   private final Map<String, ProvidedImpl> _provided        = new HashMap<>();
   @SuppressWarnings("unchecked")
   private final List<Operation>[]         _operationQueues = new List[256];

   public Dispatcher( ParticipantImpl participant ) {
      _participant = participant;
      for( int i = 0; i < _operationQueues.length; ++i ) {
         _operationQueues[i] = new LinkedList<>();
      }
   }

   @Override
   public IProvided provide( String interfaceName ) {
      final ProvidedImpl provided = new ProvidedImpl();
      _provided.put( interfaceName, provided );
      return provided;
   }

   void executeCrud( String opName, Arguments args ) {
      synchronized( _operationQueues ) {
         _operationQueues[IRequired.DEFAULT_QUEUE].add(
            new Operation( null, args, IProtocol.ICRUD_INTERFACE_NAME, opName, 0 ));
      }
   }

   boolean execute(
      String    intrfcName,
      String    opName,
      int       callId,
      Arguments arguments,
      int       queueNdx,
      CallMode  callMode ) throws IOException
   {
      final ProvidedImpl provided = _provided.get( intrfcName );
      if( provided == null ) {
         System.err.printf( "Interface '%s' isn't registered yet, call ignored\n", intrfcName );
         return false;
      }
      final IOperation operation = provided._opsInOut.get( opName );
      if( callMode == CallMode.SYNCHRONOUS ) {
         final Arguments results = operation.execute( arguments );
         if( callId > 0 ) {
            _participant.call( intrfcName, opName, results, -callId );
         }
         return true;
      }
      synchronized( _operationQueues ) {
         _operationQueues[queueNdx].add(
            new Operation( operation, arguments, intrfcName, opName, callId ));
      }
      if( callMode == CallMode.ASYNCHRONOUS_IMMEDIATE ) {
         handleRequests();
      }
      return true;
   }

   @Override
   public void handleRequests() throws IOException{
      synchronized( _operationQueues ) {
         for( final List<Operation> queue : _operationQueues ) {
            for( final Operation op : queue ) {
               final IOperation iOp = op._operation;
               if( iOp == null ) {
                  ClassID classID;
                  GUID    id;
                  switch( op._opName ) {
                  case ICRUD_INTERFACE_CREATE:
                     classID = op._arguments.get( ICRUD_INTERFACE_CLASSID );
                     _participant.create( classID, op._arguments );
                     break;
                  case ICRUD_INTERFACE_UPDATE:
                     id = op._arguments.get( ICRUD_INTERFACE_GUID );
                     _participant.update( id, op._arguments );
                     break;
                  case ICRUD_INTERFACE_DELETE:
                     id = op._arguments.get( ICRUD_INTERFACE_GUID );
                     _participant.delete( id );
                     break;
                  default:
                     System.err.printf( "Unexpected Publisher operation '%s'\n", op._opName );
                     break;
                  }
               }
               else {
                  final Arguments  results = iOp.execute( op._arguments );
                  if( op._callId > 0 ) {
                     _participant.call( op._intrfcName, op._opName, results, -op._callId );
                  }
               }
            }
            queue.clear();
         }
      }
   }

   @Override
   public IRequired require( String name ) {
      return new IRequired() {
         @Override
         public void call( String opName ) throws IOException {
            _participant.call( name, opName, null, 0 );
         }
         @Override
         public void call( String opName, Arguments arguments ) throws IOException {
            _participant.call( name, opName, arguments, 0 );
         }
         @Override
         public void call( String opName, Arguments arguments, ICallback callback ) throws IOException {
            _participant.call( name, opName, arguments, callback );
         }
      };
   }

   @Override
   public ICRUD requireCRUD( ClassID classId ) {
      return new ICRUD(){
         @Override
         public void create( Arguments how ) throws IOException {
            how.put( ICRUD_INTERFACE_CLASSID, classId );
            _participant.call( ICRUD_INTERFACE_NAME, ICRUD_INTERFACE_CREATE, how, 0 );
         }
         @Override
         public void update( Shareable what, Arguments how ) throws IOException {
            how.put( ICRUD_INTERFACE_GUID, what._id );
            _participant.call( ICRUD_INTERFACE_NAME, ICRUD_INTERFACE_UPDATE, how, 0 );
         }
         @Override
         public void delete( Shareable what ) throws IOException {
            final Arguments how = new Arguments();
            how.put( ICRUD_INTERFACE_GUID, what._id );
            _participant.call( ICRUD_INTERFACE_NAME, ICRUD_INTERFACE_DELETE, how, 0 );
         }
      };
   }
}

File: cpp/src/dcrud/Dispatcher.cpp
#include "Dispatcher.hpp"
#include "ParticipantImpl.hpp"
#include "RequiredImpl.hpp"
#include "ProvidedImpl.hpp"
#include "Operation.hpp"

#include <dcrud/ICRUD.hpp>
#include <dcrud/Shareable.hpp>

namespace dcrud {

   class CRUD : public ICRUD {
   public:

      CRUD( ParticipantImpl & participant, const ClassID & classID ) :
         _participant( participant ),
         _classID    ( classID     )
      {}

      virtual void create( const Arguments & how ) {
         Arguments args( how );
         args.put( ICRUD_INTERFACE_CLASSID, _classID );
         _participant.call( ICRUD_INTERFACE_NAME, ICRUD_INTERFACE_CREATE, &args, 0 );
      }

      virtual void update( Shareable & what, const Arguments & how ) {
         Arguments args( how );
         args.put( ICRUD_INTERFACE_GUID, what.getGUID());
         _participant.call( ICRUD_INTERFACE_NAME, ICRUD_INTERFACE_UPDATE, &args, 0 );
      }

      virtual void deleTe( Shareable & what ) {
         Arguments args;
         args.put( ICRUD_INTERFACE_GUID, what.getGUID());
         _participant.call( ICRUD_INTERFACE_NAME, ICRUD_INTERFACE_DELETE, &args, 0 );
      }

   private:

      ParticipantImpl & _participant;
      ClassID           _classID;
   };
}
using namespace dcrud;

IProvided & Dispatcher::provide( const std::string & interfaceName ) {
   os::Synchronized sync( _providedMutex );
   providedIter_t it = _provided.find( interfaceName );
   if( it == _provided.end()) {
      ProvidedImpl * provided = new ProvidedImpl();
      _provided[interfaceName] = provided;
      return *provided;
   }
   return *it->second;
}

bool Dispatcher::execute(
   const std::string & intrfcName,
   const std::string & opName,
   const Arguments &   arguments,
   int                 callId,
   byte                queueNdx,
   byte                callMode )
{
   ProvidedImpl * provided = 0;
   {
      os::Synchronized sync( _providedMutex );
      providedIter_t it = _provided.find( intrfcName );
      if( it == _provided.end()) {
         fprintf( stderr, "No provided interface named '%s' found\n", intrfcName.c_str());
         return false;
      }
      provided  = it->second;
   }
   IOperation * operation = provided->getOperation( opName );
   if( ! operation ) {
      fprintf( stderr,
         "Provided interface named '%s' is found but it's null! (internal error)\n",
         intrfcName.c_str());
      return false;
   }
   if( callMode == IRequired::SYNCHRONOUS ) {
      Arguments * results = operation->execute( _participant, arguments );
      if( callId ) {
         _participant.call( intrfcName, opName, results, -callId );
      }
   }
   else {
      {
         os::Synchronized sync(_operationQueuesMutex);
         _operationQueues[queueNdx].push_back(
            new Operation( operation, intrfcName, opName, arguments, callId ));
      }
      if( callMode == IRequired::ASYNCHRONOUS_IMMEDIATE ) {
         handleRequests();
      }
   }
   return true;
}

ICRUD & Dispatcher::requireCRUD( const ClassID & classID ) {
   return *new CRUD( this->_participant, classID );
}

void Dispatcher::executeCrud( const std::string & opName, const Arguments & arguments ) {
   os::Synchronized sync(_operationQueuesMutex);
   _operationQueues[IRequired::DEFAULT_QUEUE].push_back( new Operation( opName, arguments ));
}

IRequired & Dispatcher::require( const std::string & name ) {
   return *new RequiredImpl( name, _participant );
}

void Dispatcher::handleRequests() {
   os::Synchronized sync(_operationQueuesMutex);
   for( unsigned i = 0; i < OPERATION_QUEUE_COUNT; ++i ) {
      operations_t & queue = _operationQueues[i];
      for( operationsIter_t it = queue.begin(); it != queue.end(); ++it ) {
         Operation * op  = *it;
         try {
            IOperation * iOp = op->_operation;
            if( iOp ) {
               Arguments * results = iOp->execute( _participant, op->_arguments );
               if( op->_callId > 0 ) {
                  _participant.call( op->_intrfcName, op->_opName, results, -op->_callId );
               }
            }
            else {
               ClassID classID;
               GUID    id;
               if(   op->_opName == ICRUD_INTERFACE_CREATE
                  && op->_arguments.get( ICRUD_INTERFACE_CLASSID, classID ))
               {
                  _participant.create( classID, op->_arguments );
               }
               else if( op->_opName == ICRUD_INTERFACE_UPDATE
                  && op->_arguments.get( ICRUD_INTERFACE_GUID, id ))
               {
                  _participant.update( id, op->_arguments );
               }
               else if( op->_opName == ICRUD_INTERFACE_DELETE
                  && op->_arguments.get( ICRUD_INTERFACE_GUID, id ))
               {
                  _participant.deleTe( id );
               }
               else {
                  fprintf( stderr, "%s:%d: Unexpected Publisher operation '%s'\n",
                     __FILE__, __LINE__, op->_opName.c_str());
               }
            }
         }
         catch( const std::exception & x ) {
            fprintf( stderr, "%s\n", x.what());
         }
         delete op;
      }
      queue.clear();
   }
}
